File: C++/CPPTut/CPPTut/Polymorphism.cpp
// Tutorial from Video :
// https://www.youtube.com/watch?v=Rub-JsjMhWY 

#include <iostream> // include in out stream for cout and cin
#include <vector>   // include vectors
#include <string>   // include strings
#include <fstream>  // include file stream for file io
#include <time.h>	// unclude time function

using namespace std;	// this allows me to avoid having to type 'std::cout' everytime

class Animal {

private:
	int height;
	int weight;
	string name;

	static int numOfAnimals;	// static variable

public:
	// getters and setters
	int getHeight() { return height; }
	int getWeight() { return weight; }
	string getName() { return name; }
	void setHeight(int cm) { height = cm; }
	void setWeight(int kg) { weight = kg; }
	void setName(string animalName) { name = animalName; }

	// Prototypes
	void setAll(int, int, string);
	Animal(int, int, string);		// Constructor
	Animal();						// Constructor
	~Animal();						// Destructor
	string toString();

	static int getNumOfAnimals() { return numOfAnimals; }	// Static Function

};

int Animal::numOfAnimals = 0;	// set static variable

// decalre prototypes
void Animal::setAll(int height, int weight, string name) {
	this->height = height;
	this->weight = weight;
	this->name = name;
}

Animal::Animal(int height, int weight, string name) {
	this->height = height;
	this->weight = weight;
	this->name = name;
	numOfAnimals++;
}

Animal::Animal() {
	numOfAnimals++;
}

Animal::~Animal() {
	cout << "Animal " << this->name << " is destroyed" << endl;
}

string Animal::toString() {
	string returnStr = "Name: ";
	returnStr.append(name).append("\nHeight: ").append(to_string(height)).append(" cm");
	returnStr.append("\nWeight: ").append(to_string(weight)).append(" kg");

	return returnStr;
}

void animalExample() {
	Animal bunny;
	bunny.setHeight(87);
	bunny.setWeight(4);
	bunny.setName("Bonny");
	cout << "Name: " << bunny.getName() << endl;
	cout << "Height: " << bunny.getHeight() << " cm" << endl;
	cout << "weight: " << bunny.getWeight() << " kg" << endl << endl;

	Animal deer(125, 80, "Doe");
	cout << deer.toString() << endl << endl;
}

// Extends and Inherits
class Dog : public Animal {
private:
	string sound = "bark";
public:
	void setSound(string sound) { this->sound = sound; }
	string getSound() { return sound; }

	Dog(int, int, string, string);
	Dog() : Animal() {};
	string toString();
};

Dog::Dog(int height, int weight, string name, string sound) : Animal(height, weight, name) {
	this->sound = sound;
}

string Dog::toString() {
	string returnStr = Animal::toString();
	returnStr.append("\nSound: ").append(sound);

	return returnStr;
}

void dogExample() {
	Dog bassetHound(100, 45, "Porter", "Row");
	cout << bassetHound.toString() << endl;
}

// polymorphism
class Student {
public:
	void getFamily() { cout << "We are Students" << endl; }
	virtual void getClass() { cout << "I am a Student" << endl; }	// lets computer know this will most likely get overridden, kind of like abstract
};

class Senior : public Student {
public:
	void getClass() { cout << "I am a Senior" << endl; }
};

class MathSenior : public Senior {
public:
	void getClass() { cout << "I am a Math Senior" << endl; }
};

void whatClass(Student* student) {
	student->getClass();
}

void studentExample() {
	Student* student = new Student;		// creates a pointer that points to Object
	Senior* senior = new Senior;

	student->getClass();
	senior->getClass();
	whatClass(student);
	whatClass(senior);

	Student* seniorPtr = new Senior;	// creates a pointer that points to Object
	MathSenior stephen;
	Student* mathSeniorPtr = &stephen;	// same thing except the MathSenior Object is declared

	seniorPtr->getFamily();
	seniorPtr->getClass();
	mathSeniorPtr->getFamily();
	mathSeniorPtr->getClass();
}

/*
// all code is kept within the main ion
int main() {

	studentExample();

	return 0;	// this means we return the porgram runned perfectly fine

}
*/

File: LanguagesTutorials/PythonTut/PythonTut/Polymorphism.py
# Tutorial from Video :
# https://www.youtube.com/watch?v=N4mEzFDjqtA&t=152s

# import modules
import random
import sys
import os


class Animal:
    __name = ""  # '__' means private
    __height = 0
    __weight = 0
    __sound = 0

    # Constructor
    def __init__(self, name, height, weight, sound):
        self.__name = name
        self.__height = height
        self.__weight = weight
        self.__sound = sound

    # Getters and Setters
    def set_name(self, name):
        self.__name = name

    def set_height(self, height):
        self.__height = height

    def set_weight(self, weight):
        self.__weight = weight

    def set_sound(self, sound):
        self.__sound = sound

    def get_name(self):
        return self.__name

    def get_height(self):
        return str(self.__height)

    def get_weight(self):
        return str(self.__weight)

    def get_sound(self):
        return self.__sound

    def get_type(self):
        return "Animal Type"

    def to_string(self):
        return "Name: {}\n" \
               "Height: {} cm\n" \
               "Weight: {} kg\n" \
               "Sound: {}".format(self.__name, self.__height,
                                  self.__weight, self.__sound)


# Dog extends Animal
class Dog(Animal):
    __owner = ""

    def __init__(self, name, height, weight, sound, owner):
        self.__owner = owner
        super(Dog, self).__init__(name, height, weight, sound)

    def set_owner(self, owner):
        self.__owner = owner

    def get_owner(self):
        return self.__owner

    def get_type(self):
        return "Dog Type"

    def to_string(self):
        return super(Dog, self).to_string() + "\nOwner: " + self.__owner

    def multi_sound(self, amt_times=None):
        if amt_times is None:
            print(self.get_sound())
        else:
            print(self.get_sound() * amt_times)


def get_type(animal):
    return animal.get_type()


def animal_examples():
    snake = Animal("Slither", 55, 20, "ssss")
    print(snake.to_string())
    print(snake.get_type())
    print()

    dog = Dog("Bear", 20, 10, "ruff", "Theresa")
    print(dog.to_string())
    print(dog.get_type())
    dog.multi_sound()
    dog.multi_sound(2)
    print()

    print(get_type(snake))
    print(get_type(dog))


animal_examples()
